File: letter_combinations_of_a_phone_number.py
'''
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.

Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
'''

'''
Classic combination problem. We can solve this using either recursive or iterative way. Time complexity is O(n^k), where n is the length of digits, and k is the average number of letters for each digit.
'''

class Solution:
    # @return a list of strings, [s1, s2]
    def letterCombinations(self, digits):
        res = []
        self.letterCombinationsHelper(digits, 0, '', res)
        return res

    def letterCombinationsHelper(self, digits, index, current_string, res):
        mapping = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        if index == len(digits):
            res.append(current_string)
        else:
            for c in mapping[int(digits[index])]:
                current_string += c
                self.letterCombinationsHelper(digits, index + 1, current_string, res)
                # trace back
                current_string = current_string[:len(current_string) - 1]

    def letterCombinations1(self, digits):
        mapping = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        res = ['']
        tmp = []
        for d in digits:
            tmp = []
            for s in res:
                for c in mapping[int(d)]:
                    tmp.append(s + c)
            res = tmp
        return res


if __name__ == '__main__':
    s = Solution()
    print s.letterCombinations('23')
    print s.letterCombinations('')
    print s.letterCombinations1('23')
    print s.letterCombinations1('')

                

File: c++/letter_combinations_of_a_phone_number.cpp
/*
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.



Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
 */

/*
 * backtrack
*/

#include "helper.h"

class Solution {
public:
    map<char, string> mapping;
    Solution() {
        mapping['0'] = "";
        mapping['1'] = "";
        mapping['2'] = "abc";
        mapping['3'] = "def";
        mapping['4'] = "ghi";
        mapping['5'] = "jkl";
        mapping['6'] = "mno";
        mapping['7'] = "pqrs";
        mapping['8'] = "tuv";
        mapping['9'] = "wxyz";
    }

    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty()) {
            return res;
        }
        string cur;
        helper(res, cur, digits, 0);
        return res;
    }

    void helper(vector<string> &res, string &cur, const string &digits, int start) {
        if (cur.size() == digits.size()) {
            res.push_back(cur);
            return;
        }
        for (int i = 0;i< mapping[digits[start]].size();i++){
                cur.push_back(mapping[digits[start]][i]);
                helper(res, cur, digits, start+1);
                cur.pop_back();
        }
    }
};


int main() {
    Solution s;
    cout << s.letterCombinations("22") << endl;
    return 0;
}
